File: fsensor/src/main/java/com/tracqi/fsensor/rotation/fusion/ekf/EkfFusion.java
package com.tracqi.fsensor.rotation.fusion.ekf;

/*
 * Copyright 2024, Tracqi Technology, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Extended Kalman Filter for quaternion-based orientation estimation.
 * <p>
 * Unlike the linear Kalman filter, the EKF properly handles the non-linear
 * quaternion kinematics by linearizing the dynamics at each time step.
 * <p>
 * State: x = [q0, q1, q2, q3] (unit quaternion)
 * Process model: quaternion kinematics q_dot = 0.5 * Omega(omega) * q
 * Measurement model: expected gravity and magnetic field in body frame
 */
public class EkfFusion {

    private final double processNoiseVariance;
    private final double accelNoiseVariance;
    private final double magNoiseVariance;

    // State: quaternion [q0, q1, q2, q3]
    private double[] q = {1.0, 0.0, 0.0, 0.0};

    // Error covariance 4x4
    private double[][] P = {
            {0.1, 0, 0, 0},
            {0, 0.1, 0, 0},
            {0, 0, 0.1, 0},
            {0, 0, 0, 0.1}
    };

    /**
     * @param processNoiseVariance gyroscope noise variance (typical: 0.001)
     * @param accelNoiseVariance   accelerometer noise variance (typical: 0.1)
     * @param magNoiseVariance     magnetometer noise variance (typical: 0.5)
     */
    public EkfFusion(double processNoiseVariance, double accelNoiseVariance, double magNoiseVariance) {
        this.processNoiseVariance = processNoiseVariance;
        this.accelNoiseVariance = accelNoiseVariance;
        this.magNoiseVariance = magNoiseVariance;
    }

    public EkfFusion() {
        this(0.001, 0.1, 0.5);
    }

    /**
     * Predict step using gyroscope data.
     */
    public void predict(float[] gyro, float dt) {
        double wx = gyro[0], wy = gyro[1], wz = gyro[2];

        // State transition matrix: F = I + 0.5 * Omega(omega) * dt
        // Omega(omega) for quaternion kinematics:
        // [  0  -wx  -wy  -wz ]
        // [ wx   0    wz  -wy ]
        // [ wy  -wz   0    wx ]
        // [ wz   wy  -wx   0  ]
        double halfDt = 0.5 * dt;
        double[][] F = {
                {1.0, -halfDt * wx, -halfDt * wy, -halfDt * wz},
                {halfDt * wx, 1.0, halfDt * wz, -halfDt * wy},
                {halfDt * wy, -halfDt * wz, 1.0, halfDt * wx},
                {halfDt * wz, halfDt * wy, -halfDt * wx, 1.0}
        };

        // Propagate state: q_new = F * q
        double[] qNew = matVecMul4(F, q);

        // Normalize
        double norm = Math.sqrt(qNew[0] * qNew[0] + qNew[1] * qNew[1] + qNew[2] * qNew[2] + qNew[3] * qNew[3]);
        q[0] = qNew[0] / norm;
        q[1] = qNew[1] / norm;
        q[2] = qNew[2] / norm;
        q[3] = qNew[3] / norm;

        // Propagate covariance: P = F * P * F' + Q
        double[][] Ft = transpose4(F);
        double[][] FP = matMul4(F, P);
        double[][] FPFt = matMul4(FP, Ft);
        for (int i = 0; i < 4; i++) {
            for (int j = 0; j < 4; j++) {
                P[i][j] = FPFt[i][j] + (i == j ? processNoiseVariance * dt : 0);
            }
        }
    }

    /**
     * Correct step using accelerometer (gravity reference).
     */
    public void correctAccelerometer(float[] accel) {
        double ax = accel[0], ay = accel[1], az = accel[2];
        double normA = Math.sqrt(ax * ax + ay * ay + az * az);
        if (normA < 1e-10) return;
        ax /= normA; ay /= normA; az /= normA;

        // Expected gravity in body frame from current quaternion: R(q)^T * [0, 0, 1]
        double q0q0 = q[0] * q[0], q1q1 = q[1] * q[1], q2q2 = q[2] * q[2], q3q3 = q[3] * q[3];
        double gx_pred = 2.0 * (q[1] * q[3] - q[0] * q[2]);
        double gy_pred = 2.0 * (q[0] * q[1] + q[2] * q[3]);
        double gz_pred = q0q0 - q1q1 - q2q2 + q3q3;

        // Innovation (measurement - prediction)
        double[] innovation = {ax - gx_pred, ay - gy_pred, az - gz_pred};

        // Jacobian H (3x4): dh/dq for gravity measurement
        double[][] H = {
                {-2.0 * q[2], 2.0 * q[3], -2.0 * q[0], 2.0 * q[1]},
                {2.0 * q[1], 2.0 * q[0], 2.0 * q[3], 2.0 * q[2]},
                {2.0 * q[0], -2.0 * q[1], -2.0 * q[2], 2.0 * q[3]}
        };

        // R (3x3 measurement noise)
        double[][] R = {
                {accelNoiseVariance, 0, 0},
                {0, accelNoiseVariance, 0},
                {0, 0, accelNoiseVariance}
        };

        applyCorrection(H, R, innovation, 3);
    }

    /**
     * Correct step using magnetometer (heading reference).
     */
    public void correctMagnetometer(float[] mag) {
        double mx = mag[0], my = mag[1], mz = mag[2];
        double normM = Math.sqrt(mx * mx + my * my + mz * mz);
        if (normM < 1e-10) return;
        mx /= normM; my /= normM; mz /= normM;

        // Rotate mag to Earth frame to find reference
        double q0q0 = q[0] * q[0], q1q1 = q[1] * q[1], q2q2 = q[2] * q[2], q3q3 = q[3] * q[3];
        double hx = mx * (q0q0 + q1q1 - q2q2 - q3q3) + 2.0 * my * (q[1] * q[2] - q[0] * q[3]) + 2.0 * mz * (q[1] * q[3] + q[0] * q[2]);
        double hy = 2.0 * mx * (q[1] * q[2] + q[0] * q[3]) + my * (q0q0 - q1q1 + q2q2 - q3q3) + 2.0 * mz * (q[2] * q[3] - q[0] * q[1]);
        double bx = Math.sqrt(hx * hx + hy * hy);
        double bz = 2.0 * mx * (q[1] * q[3] - q[0] * q[2]) + 2.0 * my * (q[2] * q[3] + q[0] * q[1]) + mz * (q0q0 - q1q1 - q2q2 + q3q3);

        // Expected magnetic field in body frame
        double mx_pred = bx * (q0q0 + q1q1 - q2q2 - q3q3) + bz * 2.0 * (q[1] * q[3] - q[0] * q[2]);
        double my_pred = bx * 2.0 * (q[1] * q[2] - q[0] * q[3]) + bz * 2.0 * (q[0] * q[1] + q[2] * q[3]);
        double mz_pred = bx * 2.0 * (q[0] * q[2] + q[1] * q[3]) + bz * (q0q0 - q1q1 - q2q2 + q3q3);

        double[] innovation = {mx - mx_pred, my - my_pred, mz - mz_pred};

        // Simplified Jacobian (numerical differentiation could be used for exact)
        // Use approximate H based on gravity-like structure for heading
        double[][] H = {
                {-2.0 * q[2] * bz, 2.0 * q[3] * bz, -2.0 * q[0] * bz + 2.0 * q[2] * bx, 2.0 * q[1] * bz + 2.0 * q[3] * bx},
                {2.0 * q[1] * bz - 2.0 * q[3] * bx, 2.0 * q[0] * bz + 2.0 * q[2] * bx, 2.0 * q[3] * bz + 2.0 * q[1] * bx, -2.0 * q[0] * bx + 2.0 * q[2] * bz},
                {2.0 * q[0] * bx + 2.0 * q[2] * bz, -2.0 * q[1] * bx + 2.0 * q[3] * bz, 2.0 * q[0] * bz - 2.0 * q[2] * bx, 2.0 * q[1] * bz + 2.0 * q[3] * bx}
        };

        double[][] R = {
                {magNoiseVariance, 0, 0},
                {0, magNoiseVariance, 0},
                {0, 0, magNoiseVariance}
        };

        applyCorrection(H, R, innovation, 3);
    }

    private void applyCorrection(double[][] H, double[][] R, double[] innovation, int mDim) {
        // S = H * P * H' + R (mDim x mDim)
        double[][] Ht = new double[4][mDim];
        for (int i = 0; i < mDim; i++)
            for (int j = 0; j < 4; j++)
                Ht[j][i] = H[i][j];

        double[][] PHt = matMul(P, Ht, 4, 4, mDim);
        double[][] S = matMul(H, PHt, mDim, 4, mDim);
        for (int i = 0; i < mDim; i++)
            S[i][i] += R[i][i];

        // K = P * H' * S^-1 (4 x mDim)
        double[][] Sinv = invertSymmetric(S, mDim);
        double[][] K = matMul(PHt, Sinv, 4, mDim, mDim);

        // State update: x = x + K * innovation
        for (int i = 0; i < 4; i++) {
            double correction = 0;
            for (int j = 0; j < mDim; j++) {
                correction += K[i][j] * innovation[j];
            }
            q[i] += correction;
        }

        // Normalize quaternion
        double norm = Math.sqrt(q[0] * q[0] + q[1] * q[1] + q[2] * q[2] + q[3] * q[3]);
        q[0] /= norm; q[1] /= norm; q[2] /= norm; q[3] /= norm;

        // Covariance update: P = (I - K*H) * P (Joseph form for stability)
        double[][] KH = matMul(K, H, 4, mDim, 4);
        double[][] IminusKH = new double[4][4];
        for (int i = 0; i < 4; i++) {
            for (int j = 0; j < 4; j++) {
                IminusKH[i][j] = (i == j ? 1.0 : 0.0) - KH[i][j];
            }
        }
        P = matMul4(IminusKH, P);
    }

    public double[] getQuaternion() {
        return new double[]{q[0], q[1], q[2], q[3]};
    }

    public void reset() {
        q = new double[]{1.0, 0.0, 0.0, 0.0};
        P = new double[][]{
                {0.1, 0, 0, 0},
                {0, 0.1, 0, 0},
                {0, 0, 0.1, 0},
                {0, 0, 0, 0.1}
        };
        }

    // --- Matrix utilities for small matrices ---

    private static double[] matVecMul4(double[][] A, double[] x) {
        double[] result = new double[4];
        for (int i = 0; i < 4; i++) {
            result[i] = A[i][0] * x[0] + A[i][1] * x[1] + A[i][2] * x[2] + A[i][3] * x[3];
        }
        return result;
    }

    private static double[][] matMul4(double[][] A, double[][] B) {
        double[][] C = new double[4][4];
        for (int i = 0; i < 4; i++)
            for (int j = 0; j < 4; j++)
                for (int k = 0; k < 4; k++)
                    C[i][j] += A[i][k] * B[k][j];
        return C;
    }

    private static double[][] matMul(double[][] A, double[][] B, int m, int n, int p) {
        double[][] C = new double[m][p];
        for (int i = 0; i < m; i++)
            for (int j = 0; j < p; j++)
                for (int k = 0; k < n; k++)
                    C[i][j] += A[i][k] * B[k][j];
        return C;
    }

    private static double[][] transpose4(double[][] A) {
        double[][] T = new double[4][4];
        for (int i = 0; i < 4; i++)
            for (int j = 0; j < 4; j++)
                T[i][j] = A[j][i];
        return T;
    }

    private static double[][] invertSymmetric(double[][] A, int n) {
        // Simple Gauss-Jordan inversion for small symmetric matrices (3x3 or 4x4)
        double[][] aug = new double[n][2 * n];
        for (int i = 0; i < n; i++) {
            System.arraycopy(A[i], 0, aug[i], 0, n);
            aug[i][n + i] = 1.0;
        }

        for (int col = 0; col < n; col++) {
            // Partial pivoting
            int maxRow = col;
            for (int row = col + 1; row < n; row++) {
                if (Math.abs(aug[row][col]) > Math.abs(aug[maxRow][col])) {
                    maxRow = row;
                }
            }
            double[] temp = aug[col];
            aug[col] = aug[maxRow];
            aug[maxRow] = temp;

            double pivot = aug[col][col];
            if (Math.abs(pivot) < 1e-12) {
                pivot = 1e-12; // Regularize
            }
            for (int j = 0; j < 2 * n; j++) {
                aug[col][j] /= pivot;
            }
            for (int row = 0; row < n; row++) {
                if (row != col) {
                    double factor = aug[row][col];
                    for (int j = 0; j < 2 * n; j++) {
                        aug[row][j] -= factor * aug[col][j];
                    }
                }
            }
        }

        double[][] inv = new double[n][n];
        for (int i = 0; i < n; i++) {
            System.arraycopy(aug[i], n, inv[i], 0, n);
        }
        return inv;
    }
}

File: fsensor/src/commonMain/kotlin/com/tracqi/fsensor/fusion/ekf/EkfFusion.kt
package com.tracqi.fsensor.fusion.ekf

import com.tracqi.fsensor.fusion.FusionAlgorithm
import com.tracqi.fsensor.math.angle.Angles
import kotlin.math.sqrt

class EkfFusion(
    private val processNoiseVariance: Double = 0.001,
    private val accelNoiseVariance: Double = 0.1,
    private val magNoiseVariance: Double = 0.5
) : FusionAlgorithm {

    private var q = doubleArrayOf(1.0, 0.0, 0.0, 0.0)
    private var p = identityScaled(0.1)

    override fun update(acceleration: FloatArray, magnetic: FloatArray?, gyroscope: FloatArray, dt: Float) {
        predict(gyroscope, dt)
        correctAccelerometer(acceleration)
        if (magnetic != null) correctMagnetometer(magnetic)
    }

    fun predict(gyroscope: FloatArray, dt: Float) {
        val wx = gyroscope[0].toDouble()
        val wy = gyroscope[1].toDouble()
        val wz = gyroscope[2].toDouble()

        val omega = arrayOf(
            doubleArrayOf(0.0, -wx, -wy, -wz),
            doubleArrayOf(wx, 0.0, wz, -wy),
            doubleArrayOf(wy, -wz, 0.0, wx),
            doubleArrayOf(wz, wy, -wx, 0.0)
        )

        val f = Array(4) { i -> DoubleArray(4) { j ->
            (if (i == j) 1.0 else 0.0) + 0.5 * dt * omega[i][j]
        }}

        val newQ = matVecMul4(f, q)
        val norm = sqrt(newQ[0] * newQ[0] + newQ[1] * newQ[1] + newQ[2] * newQ[2] + newQ[3] * newQ[3])
        for (i in 0..3) q[i] = newQ[i] / norm

        val fArr = Array(4) { i -> f[i].copyOf() }
        val ft = transpose4(fArr)
        val fpft = matMul4(matMul4(fArr, p), ft)
        for (i in 0..3) for (j in 0..3) {
            p[i][j] = fpft[i][j] + (if (i == j) processNoiseVariance else 0.0)
        }
    }

    fun correctAccelerometer(acceleration: FloatArray) {
        var ax = acceleration[0].toDouble()
        var ay = acceleration[1].toDouble()
        var az = acceleration[2].toDouble()
        val normA = sqrt(ax * ax + ay * ay + az * az)
        if (normA == 0.0) return
        ax /= normA; ay /= normA; az /= normA

        val q0 = q[0]; val q1 = q[1]; val q2 = q[2]; val q3 = q[3]
        val hx = 2.0 * (q1 * q3 - q0 * q2)
        val hy = 2.0 * (q0 * q1 + q2 * q3)
        val hz = q0 * q0 - q1 * q1 - q2 * q2 + q3 * q3

        val h = arrayOf(
            doubleArrayOf(-2.0 * q2, 2.0 * q3, -2.0 * q0, 2.0 * q1),
            doubleArrayOf(2.0 * q1, 2.0 * q0, 2.0 * q3, 2.0 * q2),
            doubleArrayOf(2.0 * q0, -2.0 * q1, -2.0 * q2, 2.0 * q3)
        )

        val innovation = doubleArrayOf(ax - hx, ay - hy, az - hz)
        applyCorrection(h, innovation, accelNoiseVariance)
    }

    fun correctMagnetometer(magnetic: FloatArray) {
        var mx = magnetic[0].toDouble()
        var my = magnetic[1].toDouble()
        var mz = magnetic[2].toDouble()
        val normM = sqrt(mx * mx + my * my + mz * mz)
        if (normM == 0.0) return
        mx /= normM; my /= normM; mz /= normM

        val q0 = q[0]; val q1 = q[1]; val q2 = q[2]; val q3 = q[3]

        val rotMx = mx * (q0 * q0 + q1 * q1 - q2 * q2 - q3 * q3) + 2.0 * my * (q1 * q2 - q0 * q3) + 2.0 * mz * (q1 * q3 + q0 * q2)
        val rotMy = 2.0 * mx * (q1 * q2 + q0 * q3) + my * (q0 * q0 - q1 * q1 + q2 * q2 - q3 * q3) + 2.0 * mz * (q2 * q3 - q0 * q1)

        val bx = sqrt(rotMx * rotMx + rotMy * rotMy)
        val bz = 2.0 * mx * (q1 * q3 - q0 * q2) + 2.0 * my * (q2 * q3 + q0 * q1) + mz * (q0 * q0 - q1 * q1 - q2 * q2 + q3 * q3)

        val hx = bx * (q0 * q0 + q1 * q1 - q2 * q2 - q3 * q3) + bz * 2.0 * (q1 * q3 - q0 * q2)
        val hy = bx * 2.0 * (q1 * q2 - q0 * q3) + bz * 2.0 * (q0 * q1 + q2 * q3)
        val hz = bx * 2.0 * (q0 * q2 + q1 * q3) + bz * (q0 * q0 - q1 * q1 - q2 * q2 + q3 * q3)

        val hmx = 2.0 * (q1 * q3 - q0 * q2)
        val hmy = 2.0 * (q0 * q1 + q2 * q3)
        val hmz = q0 * q0 - q1 * q1 - q2 * q2 + q3 * q3

        val h = arrayOf(
            doubleArrayOf(-2.0 * q2, 2.0 * q3, -2.0 * q0, 2.0 * q1),
            doubleArrayOf(2.0 * q1, 2.0 * q0, 2.0 * q3, 2.0 * q2),
            doubleArrayOf(2.0 * q0, -2.0 * q1, -2.0 * q2, 2.0 * q3)
        )

        val innovation = doubleArrayOf(mx - hmx, my - hmy, mz - hmz)
        applyCorrection(h, innovation, magNoiseVariance)
    }

    private fun applyCorrection(h: Array<DoubleArray>, innovation: DoubleArray, noiseVariance: Double) {
        val ht = Array(4) { i -> DoubleArray(3) { j -> h[j][i] } }
        val pht = Array(4) { i -> DoubleArray(3) { j ->
            var sum = 0.0
            for (k in 0..3) sum += p[i][k] * ht[k][j]
            sum
        }}

        val s = Array(3) { i -> DoubleArray(3) { j ->
            var sum = 0.0
            for (k in 0..3) sum += h[i][k] * pht[k][j]
            sum + (if (i == j) noiseVariance else 0.0)
        }}

        val sInv = invert3(s) ?: return

        val k = Array(4) { i -> DoubleArray(3) { j ->
            var sum = 0.0
            for (m in 0..2) sum += pht[i][m] * sInv[m][j]
            sum
        }}

        for (i in 0..3) {
            for (j in 0..2) {
                q[i] += k[i][j] * innovation[j]
            }
        }

        val norm = sqrt(q[0] * q[0] + q[1] * q[1] + q[2] * q[2] + q[3] * q[3])
        for (i in 0..3) q[i] /= norm

        val kh = Array(4) { i -> DoubleArray(4) { j ->
            var sum = 0.0
            for (m in 0..2) sum += k[i][m] * h[m][j]
            sum
        }}

        val newP = Array(4) { i -> DoubleArray(4) { j ->
            p[i][j] - run {
                var sum = 0.0
                for (m in 0..3) sum += kh[i][m] * p[m][j]
                sum
            }
        }}
        for (i in 0..3) for (j in 0..3) p[i][j] = newP[i][j]
    }

    override fun getOrientation(): FloatArray = Angles.getAngles(q[0], q[1], q[2], q[3])
    override fun getQuaternion(): DoubleArray = q.copyOf()
    override fun reset() {
        q = doubleArrayOf(1.0, 0.0, 0.0, 0.0)
        p = identityScaled(0.1)
    }

    private fun identityScaled(scale: Double) = Array(4) { i -> DoubleArray(4) { j -> if (i == j) scale else 0.0 } }

    private fun matVecMul4(m: Array<DoubleArray>, v: DoubleArray): DoubleArray {
        return DoubleArray(4) { i -> m[i][0] * v[0] + m[i][1] * v[1] + m[i][2] * v[2] + m[i][3] * v[3] }
    }

    private fun matMul4(a: Array<DoubleArray>, b: Array<DoubleArray>): Array<DoubleArray> {
        return Array(4) { i -> DoubleArray(4) { j ->
            var sum = 0.0
            for (k in 0..3) sum += a[i][k] * b[k][j]
            sum
        }}
    }

    private fun transpose4(m: Array<DoubleArray>): Array<DoubleArray> {
        return Array(4) { i -> DoubleArray(4) { j -> m[j][i] } }
    }

    private fun invert3(m: Array<DoubleArray>): Array<DoubleArray>? {
        val a = m[0][0]; val b = m[0][1]; val c = m[0][2]
        val d = m[1][0]; val e = m[1][1]; val f = m[1][2]
        val g = m[2][0]; val h = m[2][1]; val i = m[2][2]

        val det = a * (e * i - f * h) - b * (d * i - f * g) + c * (d * h - e * g)
        if (kotlin.math.abs(det) < 1e-12) return null
        val invDet = 1.0 / det

        return arrayOf(
            doubleArrayOf((e * i - f * h) * invDet, (c * h - b * i) * invDet, (b * f - c * e) * invDet),
            doubleArrayOf((f * g - d * i) * invDet, (a * i - c * g) * invDet, (c * d - a * f) * invDet),
            doubleArrayOf((d * h - e * g) * invDet, (b * g - a * h) * invDet, (a * e - b * d) * invDet)
        )
    }
}
